ghcide/src/Development/IDE/Plugin/CodeAction/ExactPrint.hs

```diff
@@ -118,6 +119,19 @@ fixParens openDP closeDP ctxt@(L _ elems) = do
   dropHsParTy (L _ (HsParTy _ ty)) = ty
   dropHsParTy other                = other
 
+removeConstraint ::
+  -- | Predicate: Which contect to drop.
+  (LHsType GhcPs -> Bool) ->
+  LHsType GhcPs ->
+  Rewrite
+removeConstraint pred = go
+  where
+    go (L l it@HsQualTy{hst_ctxt = L l' ctxt}) = Rewrite l $ \_ -> do
+      return $ L l $ it{hst_ctxt = L l' $ filter (not . pred) ctxt}
+    go (L _ (HsParTy _ ty)) = go ty
+    go (L _ HsForAllTy{hst_body}) = go hst_body
+    go (L l other) = Rewrite l $ \_ -> return $ L l other
+
```

Comment:
Is it enough to only remove the node from AST without adjusting annotations? I didn't look into it, but I guess the unexpected space might come from the `DeltaPos` of `HsFunTy` , which determines preceding spaces of the first `a`. Printing `Anns` to see what's going on may help.

Also, `pred` is a function in Prelude, so I would change the predication another name to avoid confusion.

Reply:
This is likely the issue; getting annotations right is the single hardest part of HLS in my experience.

Reply:
This helped thanks!